astropy/table/operations.py

```diff
@@ -87,9 +88,238 @@ def _get_out_class(objs):
+    This example shows an inner join of two ``SkyCoord`` columns, taking any
+    sources within 0.2 deg to be a match.  Note the new ``sc_id`` column which
+    is added and provides a unique source identifier for the matches.
+
+      >>> from astropy.coordinates import SkyCoord
+      >>> import astropy.units as u
+      >>> from astropy.table import Table, join_skycoord
+      >>> from astropy import table
+
+      >>> sc1 = SkyCoord([0, 1, 1.1, 2], [0, 0, 0, 0], unit='deg')
+      >>> sc2 = SkyCoord([0.5, 1.05, 2.1], [0, 0, 0], unit='deg')
+
+      >>> join_func = join_skycoord(0.2 * u.deg)
+      >>> join_func(sc1, sc2)  # Associate each coordinate with unique source ID
+      (array([3, 1, 1, 2]), array([4, 1, 2]))
+
+      >>> t1 = Table([sc1], names=['sc'])
+      >>> t2 = Table([sc2], names=['sc'])
+      >>> t12 = table.join(t1, t2, join_funcs={'sc': join_skycoord(0.2 * u.deg)})
+      >>> print(t12)  # Note new `sc_id` column with the IDs from join_func()
+      sc_id   sc_1    sc_2
+            deg,deg deg,deg
+      ----- ------- --------
+          1 1.0,0.0 1.05,0.0
+          1 1.1,0.0 1.05,0.0
+          2 2.0,0.0  2.1,0.0
+
+    """
+    def join_func(sc1, sc2):
+        if distance_kind not in ('sky', '3d'):
+            raise ValueError("distance_kind must be 'sky' or '3d'")
+
+        # Call the appropriate SkyCoord method to find pairs within distance
+        sc1_search_around = getattr(sc1, f'search_around_{distance_kind}')
+        idxs2, idxs1, d2d, d3d = sc1_search_around(sc2, distance)
+
+        # Now convert that into unique identifiers for each near-pair. This is
+        # taken to be transitive, so that if points 1 and 2 are "near" and points
+        # 1 and 3 are "near", then 1, 2, and 3 are all given the same identifier.
+        # This identifier will then be used in the table join matching.
+
+        # Identifiers for each column, initialized to all zero.
+        ids1 = np.zeros(len(sc1), dtype=int)
+        ids2 = np.zeros(len(sc2), dtype=int)
+
+        # Start the identifier count at 1
+        id_ = 1
+        for idx1, idx2 in zip(idxs1, idxs2):
+            # If this col1 point is previously identified then set corresponding
+            # col2 point to same identifier.  Likewise for col2 and col1.
+            if ids1[idx1] > 0:
+                ids2[idx2] = ids1[idx1]
+            elif ids2[idx2] > 0:
+                ids1[idx1] = ids2[idx2]
+            else:
+                # Not yet seen so set identifer for col1 and col2
+                ids1[idx1] = id_
+                ids2[idx2] = id_
+                id_ += 1
+
+        # Fill in unique identifiers for points with no near neighbor
+        for ids in (ids1, ids2):
+            for idx in np.flatnonzero(ids == 0):
+                ids[idx] = id_
+                id_ += 1
+
+        # End of enclosure join_func()
+        return ids1, ids2
+
+    return join_func
+
+
+def join_distance(distance, kdtree_args=None, query_args=None):
```

Comment:
Might it make sense to re-purpose people's experience with `isclose` and have an `atol` and `rtol`? (in which case this might become `join_close`).

Reply:
This would need to be a completely new function since this is explicitly applicable to N-dimensional points and uses KDTree to find the neighbors.  For N-dim there is no concept of `rtol`.

For the 1-d case, I actually don't know how to do a cross match with `atol` and `rtol` without just doing the N^2 cross-match.

There is another problem I thought of, which is that this whole concept is based on the cross-matching being symmetric.  If point A is "close" to point B, then it is required that B is "close" to A.  That is true for a distance (`atol`), but breaks down for `rtol`.
```
In [7]: np.isclose(1, 0, atol=0.8, rtol=0.8)                                                                                                                  
Out[7]: False

In [8]: np.isclose(0, 1, atol=0.8, rtol=0.8)                                                                                                                  
Out[8]: True
```

Reply:
Oh, yes, my brain was not quite there with the 2-d case.  Let's just stick with `distance` which is nicely defined. (and, yes, `isclose` is not symmetric unlike the `math` equivalent; https://github.com/numpy/numpy/issues/10161)